### ISSUE TITLE
way to compile only BLAS functions for a given precision ?

### ISSUE THREAD

**Issue by franko** [xianyi/OpenBLAS#2588]
Normally when compiling OpenBLAS all the functions for single and double precision floating point are generated.

I would like to know if it is possible to compile only the functions for single or double precision. I didn't find any mention about this in the documentation but it seems that Netlib's LAPACK allow to choose the precision when building.

Thanks in advance for any help.

**Comment by martin-frbg** [comment]
This is currently not supported in OpenBLAS - our cmake build files do copy the BUILD_SINGLE etc. from Netlib's LAPACK but they are only used in building LAPACK. The plain Makefiles do not support these options at all right now, although they mostly have the BLAS functions grouped by precision. #2560 is somewhat related

**Comment by franko** [comment]
@martin-frbg Thank you for the message. So not supported and it could be implemented with the Makefiles but require some works.

Probably there are not a lot of requests in this sense.

**Comment by martin-frbg** [comment]
I do not remember seeing such a request, but as I mentioned something similar just came up. And if  the reference implementation already supports it, perhaps it would make sense to add it here as well.

**Comment by brada4** [comment]
Only place I know this is done is R where only double precision parts of reference BLAS are compiled from large concatenated blas.f and lapack.f files. Resulting library then can be safely replaced with any more complete BLAS (and LAPACK) implementation. In practice though, even some R CRAN modules expect libRblas library to export single-precision or even CBLAS symbols.

**Comment by martin-frbg** [comment]
Picking them apart turns out to be a bit tricky, as some "real" functions like SSCAL are called from "complex" code.

**Comment by martin-frbg** [comment]
Untwined the gmake build (except CBLAS/LAPACKE for now) so far but not cmake yet. (Conversely, the support for this in Reference-LAPACK appears to be cmake only)

**Comment by martin-frbg** [comment]
Should work for CMAKE builds after #2833 (although with BUILD_COMPLEX one currently has to specify BUILD_SINGLE as well, and same for COMPLEX16/DOUBLE). Unfortunately need to redo the changes for gmake builds

**Comment by franko** [comment]
@martin-frbg Thank you very much, that's great. Unfortunately I cannot test right now so feel free to make your own tests and close the issue if you think it is ok.

**Comment by franko** [comment]
Hi @martin-frbg, just tested using the cmake command:

```sh
cmake -G Ninja -DDYNAMIC_ARCH=TRUE -DBUILD_COMPLEX16=TRUE -DBUILD_DOUBLE=TRUE ..
```

and I got a metric ton of undefined references at link time like:

```
lib/libopenblas.a(setparam_COOPERLAKE.c.o):(.data.rel+0xf58): undefined reference to `cimatcopy_k_rnc_COOPERLAKE'
lib/libopenblas.a(setparam_COOPERLAKE.c.o):(.data.rel+0xf60): undefined reference to `cimatcopy_k_rtc_COOPERLAKE'
lib/libopenblas.a(setparam_COOPERLAKE.c.o):(.data.rel+0xfa8): undefined reference to `sgeadd_k_COOPERLAKE'
lib/libopenblas.a(setparam_COOPERLAKE.c.o):(.data.rel+0xfb8): undefined reference to `cgeadd_k_COOPERLAKE'
collect2: error: ld returned 1 exit status
```

so it is trying to link with the single precision functions but they are not there.

**Comment by martin-frbg** [comment]
Indeed I had not yet tried with DYNAMIC_ARCH, sorry - IIRC it was just a matter of adding `ifdef BUILD_SINGLE` etc around the respective sections in kernel/setparam-ref.c

**Comment by martin-frbg** [comment]
Unfortunately changing just the one file for dynamic arch support is not sufficient and any related changes will require completing the whole set of changes for both cmake and gmake build systems. I will try to get to that this weekend

**Comment by franko** [comment]
Thank you @martin-frbg, if you want some help you don't hesitate to ask, I don't have a lot of spare time but I may be able to help you.

**Comment by franko** [comment]
Hello @martin-frbg , can I ask why you closed the issue ?

**Comment by martin-frbg** [comment]
Implemented in 0.3.11 (which would have been released by now if I had not found another blocker)

**Comment by franko** [comment]
Thank you!
